packages/client-sdk-nodejs/src/internal/grpc/client-timeout-interceptor.ts

```diff
@@ -1,14 +1,123 @@
-      options.deadline = deadline;
-    }
-    return new InterceptingCall(nextCall(options));
-  };
+  if (
+    retryStrategy !== undefined &&
+    retryStrategy instanceof DefaultStorageRetryStrategy
+  ) {
+    const responseDataReceivedTimeoutMs =
+      retryStrategy.getResponseDataReceivedTimeoutMillis();
+    return new RetryUntilTimeoutInterceptor(
+      loggerFactory ?? new DefaultMomentoLoggerFactory(),
+      overallRequestTimeoutMs,
+      responseDataReceivedTimeoutMs
+    ).createTimeoutInterceptor();
+  }
+  return new BasicTimeoutInterceptor(
+    overallRequestTimeoutMs
+  ).createTimeoutInterceptor();
 };
+
+export class BasicTimeoutInterceptor {
+  private readonly overallRequestTimeoutMs: number;
+
+  constructor(overallRequestTimeoutMs: number) {
+    this.overallRequestTimeoutMs = overallRequestTimeoutMs;
+  }
+
+  public createTimeoutInterceptor(): Interceptor {
+    return (options, nextCall) => {
+      if (!options.deadline) {
+        const deadline = new Date(Date.now());
+        deadline.setMilliseconds(
+          deadline.getMilliseconds() + this.overallRequestTimeoutMs
+        );
+        options.deadline = deadline;
+      }
+      return new InterceptingCall(nextCall(options));
+    };
+  }
+}
+
+export class RetryUntilTimeoutInterceptor {
+  private readonly logger: MomentoLogger;
+  private readonly responseDataReceivedTimeoutMs: number;
+  private readonly overallRequestTimeoutMs: number;
+  private overallDeadline?: Date;
+
+  constructor(
+    loggerFactory: MomentoLoggerFactory,
+    overallRequestTimeoutMs: number,
+    responseDataReceivedTimeoutMs: number
+  ) {
+    this.logger = loggerFactory.getLogger(this);
+    this.responseDataReceivedTimeoutMs = responseDataReceivedTimeoutMs;
+    this.overallRequestTimeoutMs = overallRequestTimeoutMs;
+  }
+
+  public createTimeoutInterceptor(): Interceptor {
+    return (options, nextCall) => {
+      // Replace default timeout with the desired overall timeout
+      // on the first time through and set the first incremental timeout
+      if (this.overallDeadline === undefined) {
+        const newDate = new Date(Date.now());
+        newDate.setMilliseconds(
+          newDate.getMilliseconds() + this.overallRequestTimeoutMs
+        );
+        this.overallDeadline = newDate;
+
+        const deadline = new Date(Date.now());
+        deadline.setMilliseconds(
+          deadline.getMilliseconds() + this.responseDataReceivedTimeoutMs
+        );
+        options.deadline = deadline;
+        this.logger.debug(`new deadline set to ${options.deadline.valueOf()}`);
+        return new InterceptingCall(nextCall(options));
+      }
+
+      const receivedDeadline = options.deadline?.valueOf() || 0;
+      this.logger.debug(
+        `intercepting call with options.deadline ${receivedDeadline}`
+      );
+
+      // Reset incremental deadline only if it has been reached
+      if (receivedDeadline < Date.now()) {
```

Comment:
this is probably okay, but a little different than what I'd been imagining. I'd been imagining that if we had hit a timeout then we would have gotten a response status of DEADLINE_EXCEEDED, and that we could check that to decide whether to reset the deadline.

I haven't thought that through completely, it's possible there are cases where your approach is better and handles something that the one I am suggesting would not, but it's worth thinking about.